tracer/src/Datadog.Trace/TraceProcessors/TraceUtil.cs

```diff
@@ -0,0 +1,344 @@
+        }
+
+        // https://github.com/DataDog/datadog-agent/blob/eac2327c5574da7f225f9ef0f89eaeb05ed10382/pkg/trace/traceutil/normalize.go#L213-L216
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool IsAlpha(char c)
+        {
+            return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z');
+        }
+
+        // https://github.com/DataDog/datadog-agent/blob/eac2327c5574da7f225f9ef0f89eaeb05ed10382/pkg/trace/traceutil/normalize.go#L218-L221
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool IsAlphaNum(char c)
+        {
+            return IsAlpha(c) || (c >= '0' && c <= '9');
+        }
+
+        // https://github.com/DataDog/datadog-agent/blob/eac2327c5574da7f225f9ef0f89eaeb05ed10382/pkg/trace/traceutil/normalize.go#L223-L274
+        public static string NormMetricNameParse(string name, int limit)
+        {
+            if (name == string.Empty || Encoding.GetByteCount(name) > limit)
+            {
+                return null;
+            }
+
+            var sb = StringBuilderCache.Acquire(name.Length);
+            int i = 0;
+
+            // skip non-alphabetic characters
+            for (; i < name.Length && !IsAlpha(name[i]); i++) { }
+
+            // if there were no alphabetic characters it wasn't valid
+            if (i == name.Length)
+            {
+                return null;
+            }
+
+            for (; i < name.Length; i++)
+            {
+                char c = name[i];
+
+                if (IsAlphaNum(c))
+                {
+                    sb.Append(c);
+                }
+                else if (c == '.')
+                {
+                    // we skipped all non-alpha chars up front so we have seen at least one
+
+                    // overwrite underscores that happen before periods
+                    if (sb[sb.Length - 1] == '_')
+                    {
+                        sb[sb.Length - 1] = '.';
+                    }
+                    else
+                    {
+                        sb.Append('.');
+                    }
+                }
+                else
+                {
+                    // we skipped all non-alpha chars up front so we have seen at least one
+
+                    // no double underscores, no underscores after periods
+                    switch (sb[sb.Length - 1])
+                    {
+                        case '.':
+                        case '_':
+                        default:
+                            sb.Append('_');
+                            break;
+                    }
+                }
+            }
+
+            if (sb[sb.Length - 1] == '_')
+            {
+                sb.Remove(sb.Length - 1, 1);
+            }
+
+            return StringBuilderCache.GetStringAndRelease(sb);
+        }
+
+        // We cannot use Span<T> and ArraySegment<T> is readonly :(
+        private readonly ref struct ArraySlice<T>
```

Comment:
If you think this could be useful, consider moving it into its own file under `Datadog.Trace/Util`.

Side note: I was thinking about doing something like `StringSegment` in my current branch to avoid some `string.Substring()` calls.

https://github.com/dotnet/runtime/blob/03494f3dcaefc3dcdbb281a19db7633c750b921c/src/libraries/Microsoft.Extensions.Primitives/src/StringSegment.cs

Reply:
Moved in [5c7da39](https://github.com/DataDog/dd-trace-dotnet/pull/2402/commits/5c7da39d240919dc399bd1461bf98ada67a6ec02)